pkg/parser/parser.go

```diff
@@ -330,10 +324,25 @@ func getServiceEndpoints(log logrus.FieldLogger, s store.Storer, svc corev1.Serv
 		"service_port":      servicePort,
 	})
 
-	endpoints = getEndpoints(log, &svc, servicePort, corev1.ProtocolTCP, s.GetEndpointsForService)
+	// in theory a Service could have multiple port protocols, we need to ensure we gather
+	// endpoints based on all the protocols the service is configured for. We always check
+	// for TCP as this is the default protocol for service ports.
+	protocols := map[corev1.Protocol]bool{corev1.ProtocolTCP: true}
```

Comment:
> The _We always check for TCP as this is the default protocol for service ports._ feels unnecessary because the API spec guarantees that the field of `Service` be filled in (on admission).

Yes, but the protocol is a different story.

> Furthermore, the "if it ain't set, assume TCP" philosophy is IMO dangerous because this can lead to harmful assumptions that the protocol we're proxying is TCP if we don't know what it is. 
> Suggestion: drop the `ProtocolTCP: true` and let the loop over ports find it.

The first time I started looking at this issue I did exactly what you suggest, however I started having growing concerns about the different in behavior from V1 in an area of the code which I would argue could use considerable amounts more testing (e.g. there's no test I can see right now that seems to exercise and illustrate why the decision to always default to TCP no matter what existing previously, let me know if I've overlooked something).

Making this default to TCP as a way to minimize behavioral changes for code paths that come through here (practically everything does really) was very intentional and I would like to leave the base logic alone if that's something you feel is justified given my above explanation.

Reply:
It seems that we think the same way and I feel convinced that making these changes to the existing TCP code paths can be too risky of regressions.

It also seems that we agree that implicitly assuming TCP here is technical debt.

Would it be a reasonable compromise to create an issue (to get rid of the implied TCP protocol in each service) and mark the addition of TCP as debt with a TODO and link to the issue?

Reply:
Seems as though we are aligned I've created https://github.com/Kong/kubernetes-ingress-controller/issues/1429 and I'm adding it as a TODO item b922d34e2e4224ef8160c99710fd83ab1242cb9e

:+1: